lib/auditd/auditd_linux.go

```diff
@@ -0,0 +1,288 @@
+	BacklogLimit          uint32 /* waiting messages limit */
+	Lost                  uint32 /* messages lost */
+	Backlog               uint32 /* messages waiting in queue */
+	Version               uint32 /* audit api version number or feature bitmap */
+	BacklogWaitTime       uint32 /* message queue wait timeout */
+	BacklogWaitTimeActual uint32 /* message queue wait timeout */
+}
+
+// IsLoginUIDSet returns true if login UID is set, false otherwise.
+func IsLoginUIDSet() bool {
+	if !hasCapabilities() {
+		// Current process doesn't have system permissions to talk to auditd.
+		return false
+	}
+
+	client := NewClient(Message{})
+	if client.connect() != nil {
+		// connect returns an error when auditd is disabled,
+		// or when we were not able to talk to it.
+		return false
+	}
+
+	loginuid, err := getSelfLoginUID()
+	if err != nil {
+		log.WithError(err).Debug("failed to read login UID")
+		return false
+	}
+
+	// if value is not set, logind PAM module will set it to the correct value
+	// after fork. 4294967295 is -1 converted to uint32.
+	return loginuid != 4294967295
+}
+
+func getSelfLoginUID() (int64, error) {
+	data, err := os.ReadFile("/proc/self/loginuid")
+	if err != nil {
+		return 0, trace.ConvertSystemError(err)
+	}
+
+	loginuid, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	return loginuid, nil
+}
+
+// SendEvent sends a single auditd event. Each request create a new netlink connection.
+// This function returns no error if it runs with no root permissions.
+func SendEvent(event EventType, result ResultType, msg Message) error {
+	if !hasCapabilities() {
+		// Disable auditd when not running as a root.
+		return nil
+	}
+
+	msg.SetDefaults()
+
+	client := NewClient(msg)
+	defer func() {
+		err := client.Close()
+		if err != nil {
+			log.WithError(err).Error("failed to close auditd client")
+		}
+	}()
+
+	if err := client.SendMsg(event, result); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (c *Client) connect() error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if !c.enabled && c.conn != nil {
+		return ErrAuditdDisabled
+	}
+
+	conn, err := c.dial(syscall.NETLINK_AUDIT, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	c.conn = conn
+
+	status, err := getAuditStatus(c.conn)
+	if err != nil {
+		return trace.Errorf("failed to get auditd status: %v", trace.ConvertSystemError(err))
+	}
+
+	c.enabled = status.Enabled == 1
+
+	if status.Enabled != 1 {
+		return ErrAuditdDisabled
+	}
+
+	return nil
+}
+
+// NewClient creates a new auditd client. Client is not connected when is returned.
+func NewClient(msg Message) *Client {
+	msg.SetDefaults()
+
+	execName, err := os.Executable()
+	if err != nil {
+		log.WithError(err).Warn("failed to get executable name")
+		execName = "?"
+	}
+
+	// Teleport never tries to get the hostname name.
+	// Let's mimic the sshd behavior.
+	const hostname = "?"
```

Comment:
What is the reason for using `"?"`? Can we set it to a constant, like `AuditdUnknown;= "?"`?